File: hey.py
import sys
import os
import google.generativeai as genai
from google.api_core import exceptions
import requests

os.environ["GRPC_ENABLE_FORK_SUPPORT"] = "0"
os.environ["GRPC_VERBOSITY"] = "NONE"
os.environ["GRPC_POLL_STRATEGY"] = "poll"

def check_internet():
    try:
        requests.get("https://www.google.com", timeout=5)
        return True
    except requests.ConnectionError:
        return False

def exception_handler(func):
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except exceptions.InvalidArgument:
            print("❌ Invalid API Key or Argument. Deleting config file.")
            if os.path.exists("config.txt"):
                os.remove("config.txt")
            sys.exit(1)
        except exceptions.NotFound:
            print("❌ Invalid model name. Deleting config file.")
            if os.path.exists("config.txt"):
                os.remove("config.txt")
            sys.exit(1)
        except exceptions.ResourceExhausted:
            print("⏳ Rate limit exceeded. Try again later.")
            sys.exit(1)
        except exceptions.GoogleAPIError as e:
            print(f"❌ API error: {e}")
            sys.exit(1)
        except Exception as e:
            print(f"⚠️ Unexpected error: {e}")
            sys.exit(1)
    return wrapper

def initialize_api():
    if not check_internet():
        print("❌ No internet connection. Please check your network and try again.")
        sys.exit(1)
    
    config_file = "config.txt"
    api_key, model = None, None
    
    if os.path.exists(config_file):
        try:
            with open(config_file, "r") as file:
                lines = file.read().splitlines()
                config = dict(line.split("=", 1) for line in lines if "=" in line)
                api_key = config.get("api")
                model = config.get("model")
        except Exception:
            print("⚠️ Error reading config file. Re-enter details.")
    
    if not api_key or not model:
        print("🔑 API key and model required.")
        api_key = input("Enter API Key: ").strip()
        model = input("Enter Model: ").strip()
        with open(config_file, "w") as file:
            file.write(f"api={api_key}\nmodel={model}")
    
    try:
        genai.configure(api_key=api_key)
        return model
    except Exception as e:
        print(f"❌ Failed to initialize API: {e}")
        sys.exit(1)

@exception_handler
def chat_with_gemini(query, instruction, model):
    model_instance = genai.GenerativeModel(model)
    history = model_instance.start_chat()
    history.send_message(instruction)
    
    response = history.send_message(query, stream=True)
    for chunk in response:
        words = chunk.text.split()
        for word in words:
            print(word, end=" ", flush=True)
    print("\n")

@exception_handler
def interactive_mode(instruction, model):
    os.system("cls" if os.name == "nt" else "clear")
    print("💬 Interactive mode enabled (Type 'exit' to quit)\n")
    
    model_instance = genai.GenerativeModel(model)
    history = model_instance.start_chat()
    history.send_message(instruction)
    
    while True:
        user_input = input("You: ").strip()
        if user_input.lower() in ["exit", "quit"]:
            print("👋 Goodbye!")
            break
        response = history.send_message(user_input, stream=True)
        print("AI:", end=" ", flush=True)
        for chunk in response:
            words = chunk.text.split()
            for word in words:
                print(word, end=" ", flush=True)
        print("\n")

def main():
    model = initialize_api()
    args = sys.argv[1:]
    
    interactive = False
    instruction = "You are an AI assistant."
    force_concise = None
    
    if args and args[0].startswith("-"):
        flags = args[0][1:]
        query_parts = args[1:]

        interactive = "i" in flags
        if "t" in flags:
            instruction = "Act as a Linux/CMD/Powershell command expert and help me fix commands. Only provide the command with a minimal comment, as your output will go directly to the terminal."
        else:
            if interactive:
                force_concise = "f" in flags
            else:
                force_concise = "n" not in flags
        
        if force_concise:
            instruction = f"Stay to the point and say less."
    else:
        query_parts = args

    if interactive or not query_parts:
        interactive_mode(instruction, model)
        return
    
    if instruction == "You are an AI assistant.":
        instruction = f"Stay to the point and say less."

    query = " ".join(query_parts)
    chat_with_gemini(query, instruction, model)

if __name__ == "__main__":
    main()

File: hey.go
package main

import (
    "bufio"
    "context"
    "fmt"
    "log"
    "os"
    "strings"

    "github.com/google/generative-ai-go/genai"
    "google.golang.org/api/option"
)

const endOfStreamErr = "no more items in iterator"

// func checkInternet() bool {
//     client := &http.Client{
//         Timeout: 5 * time.Second,
//     }
//     resp, err := client.Get("https://www.google.com")
//     if err != nil {
//         return false
//     }
//     resp.Body.Close()
//     return true
// }

func initializeAPI() (string, string) {
    // if !checkInternet() {
    //     log.Fatal("❌ No internet connection. Please check your network and try again.")
    // }

    configFile := "config.txt"
    var apiKey, model string

    if file, err := os.Open(configFile); err == nil {
        scanner := bufio.NewScanner(file)
        config := make(map[string]string)
        for scanner.Scan() {
            parts := strings.SplitN(scanner.Text(), "=", 2)
            if len(parts) == 2 {
                config[parts[0]] = parts[1]
            }
        }
        file.Close()
        apiKey, model = config["api"], config["model"]
    }

    if apiKey == "" || model == "" {
        fmt.Println("🔑 API key and model required.")
        fmt.Print("Enter API Key: ")
        scanner := bufio.NewScanner(os.Stdin)
        scanner.Scan()
        apiKey = strings.TrimSpace(scanner.Text())

        fmt.Print("Enter Model: ")
        scanner.Scan()
        model = strings.TrimSpace(scanner.Text())

        file, err := os.Create(configFile)
        if err == nil {
            file.WriteString(fmt.Sprintf("api=%s\nmodel=%s\n", apiKey, model))
            file.Close()
        }
    }

    return apiKey, model
}

func chatWithGemini(ctx context.Context, client *genai.Client, query, instruction, model string) {
    gm := client.GenerativeModel(model)
    content := genai.Text(instruction + " " + query)
    iter := gm.GenerateContentStream(ctx, content)
    if iter == nil {
        log.Fatalf("Error initiating stream")
    }

    for {
        resp, err := iter.Next()
        if err != nil {
            if strings.Contains(err.Error(), endOfStreamErr) {
                break
            }
            log.Fatalf("Error streaming message: %v", err)
        }
        for _, part := range resp.Candidates[0].Content.Parts {
            fmt.Printf("%v ", part)
            os.Stdout.Sync()
        }
    }
    fmt.Println()
}

func interactiveMode(ctx context.Context, client *genai.Client, instruction, model string) {
    fmt.Println("💬 Interactive mode enabled (Type 'exit' to quit)")

    gm := client.GenerativeModel(model)
    session := gm.StartChat()
    session.SendMessage(ctx, genai.Text(instruction))

    scanner := bufio.NewScanner(os.Stdin)
    for {
        fmt.Print("You: ")
        scanner.Scan()
        userInput := strings.TrimSpace(scanner.Text())
        if userInput == "exit" || userInput == "quit" {
            fmt.Println("👋 Goodbye!")
            break
        }
        iter := session.SendMessageStream(ctx, genai.Text(userInput))
        if iter == nil {
            log.Fatalf("Error initiating stream")
        }
        fmt.Print("AI: ")
        for {
            resp, err := iter.Next()
            if err != nil {
                if strings.Contains(err.Error(), endOfStreamErr) {
                    break
                }
                log.Fatalf("Error streaming message: %v", err)
            }
            for _, part := range resp.Candidates[0].Content.Parts {
                fmt.Printf("%v ", part)
                os.Stdout.Sync()
            }
        }
        fmt.Println()
    }
}

func main() {
    ctx := context.Background()
    apiKey, model := initializeAPI()

    client, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
    if err != nil {
        log.Fatalf("Error creating client: %v", err)
    }
    defer client.Close()

    args := os.Args[1:]
    interactive := false
    instruction := "You are an AI assistant."
    var forceConcise *bool

    if len(args) > 0 && strings.HasPrefix(args[0], "-") {
        flags := args[0][1:]
        queryParts := args[1:]

        interactive = strings.Contains(flags, "i")
        if strings.Contains(flags, "t") {
            instruction = "Act as a Linux/CMD/Powershell command expert and help me fix commands. Only provide the command with a minimal comment, as your output will go directly to the terminal."
        } else {
            if interactive {
                f := strings.Contains(flags, "f")
                forceConcise = &f
            } else {
                n := !strings.Contains(flags, "n")
                forceConcise = &n
            }
        }

        if forceConcise != nil && *forceConcise {
            instruction = "Stay to the point and say less."
        }

        if interactive || len(queryParts) == 0 {
            interactiveMode(ctx, client, instruction, model)
            return
        }

        query := strings.Join(queryParts, " ")
        chatWithGemini(ctx, client, query, instruction, model)
        return
    }

    if len(args) == 0 {
        interactiveMode(ctx, client, instruction, model)
    } else {
        query := strings.Join(args, " ")
        chatWithGemini(ctx, client, query, "Stay to the point and say less.", model)
	   }
}